h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -170,6 +170,29 @@ private static int ensureBuilderIndex(String urlName) {
     return mb;
   }
 
+  /**
+   * Method to create deep copy of ModelBuilder so that it could be trained independently without affecting original builder.
+   */
+  public <B extends ModelBuilder> B makeCopy() {
```

Comment:
Not sure what would be the best name for the method. This is not deepCloning as I did not care about all possible copies of objects that might be referencing by ModelBuilder's instance. It is not a shallow cloning as well as I do deep copy of the frames.

Reply:
@deil87 can you please provide a concrete example why you would need a copy of the `ModelBuilder` itself?
Why not simply cloning the parameters?
Why deep copying the frames? looks like this could be a quite expensive operation.

Reply:
@seb-h2o please read javadoc for `trainModelOnH2ONode` method of MB. 
```
* Warning: the nature of this method prohibits further use of this instance of the model builder after the method
   *          is called.
   *
   * This is intended to reduce training time in client-mode setups, it pushes all computation to a regular H2O node
   * and avoid exchanging data between client and H2O cluster. This also lowers requirements on the H2O client node.
```
Training process leaves some intermediate data in the original frames which makes it dangerous to reuse them. I was relying on the comment's warning when decided to make a copy of the MB itself.  `ModelBuilder` is quite mutable structure.... there are `Countdowns` as you might know.  Also we need a job with new key to be able to submit training with different TE hyperparameters. 

I think that the most expensive operation is a copying of the data. We can probably come up with non-disruptive training method but this on the other hand will impact speed of training(again according to a javadoc). If you have any suggestions I would be glad to discuss them.

Reply:
@deil87 My point is that you can create a new `ModelBuilder` just with the `Model.Parameters` object, and you can create as many as you want from the same parameters.
Now given that we know that for now, training may mutate those parameters, you can just clone them, and have a new `ModelBuilder` out if it.
I just don't understand the necessity of another method for that (`new ModelBuilder(params.clone())`.
But here, you also deep clone training and validation frame, which should not be modified by MB afaik, so I'm just wondering about the reason behind this.

Reply:
@seb-h2o  There would not be such a Warning in a comment  to the method `trainModelOnH2ONode` if everything was that simple. Please take a look at screenshot that shows `size=-1` for the training frame right after training is finished and also pay attention to the names of the columns.  <img width="1352" alt="Screenshot 2019-09-23 at 23 06 17" src="https://user-images.githubusercontent.com/1162131/65462927-1194db80-de57-11e9-987b-79083b7d8e9a.png"> . 

Frames are being modified by the training process. Again, please see the comment to the method - it is done to `...avoid exchanging data between client and H2O cluster.` And as frames are being modified at some point you will need to deep copy them from parameters. PR could be rewritten from perspective where we don't clone ModelBuilder but clone parameters with deepCopy of the frames - that is actually what I'm doing but I put the code inside ModelBuilder.

Reply:
`trainModelOnH2ONode` is unrelated IMHO, client mode refers to sparkling water client mode in this context.

Reply:
Thank you guys. Was confused by the javadoc. Maybe it does not make sense to introduce this method after all. 
@seb-h2o params stay intact and looks like we don't even need to clone them. Did not find constructor like in your comment as ModelBuilder is an abstract class, but I guess it was just a pseudocode to express an idea.